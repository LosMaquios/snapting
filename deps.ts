export * as path from "https://deno.land/std@0.74.0/path/mod.ts";
export * as asserts from "https://deno.land/std@0.74.0/testing/asserts.ts";
