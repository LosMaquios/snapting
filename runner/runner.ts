import { SnapshotManager } from "../manager/manager.ts";
import type { SnapshotCase } from "../case/case.ts";

const managerCache = new Map<string, SnapshotManager>();

let currentSnapshotCase: SnapshotCase | null;

export function getCurrentSnapshotCase() {
  return currentSnapshotCase;
}

export function createSnapshotRunner(importURL: string) {
  return function snapshotRunner(...args: any[]) {
    let testDefinition: Deno.TestDefinition = args[0];

    if (typeof args[0] !== "object") {
      testDefinition = {
        name: args[0],
        fn: args[1],
      };
    }

    const { fn } = testDefinition;
    let snapshotManager = managerCache.get(importURL)!;

    if (!snapshotManager) {
      managerCache.set(
        importURL,
        snapshotManager = new SnapshotManager(importURL),
      );
    }

    // Patch test fn
    testDefinition.fn = async (...args) => {
      await snapshotManager.resolveSnapshotFile();

      currentSnapshotCase = snapshotManager.getCase(testDefinition.name);
      await fn(...args);

      currentSnapshotCase.validatePendingAssertions();
      currentSnapshotCase = null;
    };

    // Attach patched test fn
    Deno.test(testDefinition);
  } as typeof Deno["test"];
}
