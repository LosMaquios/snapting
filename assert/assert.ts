import { getCurrentSnapshotCase } from "../runner/runner.ts";

export function assertSnapshot(expected: unknown, message = "") {
  const currentSnapshotCase = getCurrentSnapshotCase();

  if (!currentSnapshotCase) {
    throw new Error("`assertSnapshot` must be called within a test runner");
  }

  return currentSnapshotCase.assert(expected, message);
}
