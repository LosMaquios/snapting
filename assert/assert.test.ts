import { asserts } from "../deps.ts";
import { assertSnapshot } from "../assert/assert.ts";
import { createSnapshotRunner } from "../runner/runner.ts";

const test = createSnapshotRunner(import.meta.url);

test("should work with primitive types", () => {
  // Assert: number
  assertSnapshot(123, "works with integers");
  assertSnapshot(123.456, "works with floats");
  // TODO: assertSnapshot(123n ** 4567890n, "works with bigints");

  // Assert: string
  assertSnapshot("some string", "works with strings");

  // Assert: boolean
  assertSnapshot(true, "works with true");
  assertSnapshot(false, "works with false");

  // Assert: null
  assertSnapshot(null);
});

test("should work with complex types", () => {
  // Assert: object
  assertSnapshot({
    int: 123,
    float: 123.456,
    string: "some string",
    truthy: true,
    falsy: false,
    nullish: null,
    object: {
      key: "value",
    },
    array: [
      "element",
    ],
  }, "works with object");

  // Assert: Array
  assertSnapshot([
    123,
    123.456,
    "some string",
    true,
    false,
    null,
    {
      key: "value",
    },
    [
      "element",
    ],
  ], "works with arrays");

  // TODO: assertSnapshot(function fn() {}, "works with function")
  // TODO: assertSnapshot(/[a-z]+/i, "works with regex")
});

Deno.test("should throw on invalid `assertSnapshot` call", () => {
  try {
    assertSnapshot("Something wrong!");
    asserts.assert(false);
  } catch (error) {
    asserts.assert(error instanceof Error);
    asserts.assert(
      error.message === "`assertSnapshot` must be called within a test runner",
    );
  }
});
