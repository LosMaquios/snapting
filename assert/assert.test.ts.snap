export const assertions = {};

assertions["should work with primitive types 0"] = 123;

assertions["should work with primitive types 1"] = 123.456;

assertions["should work with primitive types 2"] = "some string";

assertions["should work with primitive types 3"] = true;

assertions["should work with primitive types 4"] = false;

assertions["should work with primitive types 5"] = null;

assertions["should work with complex types 0"] = {
  "int": 123,
  "float": 123.456,
  "string": "some string",
  "truthy": true,
  "falsy": false,
  "nullish": null,
  "object": {
    "key": "value"
  },
  "array": [
    "element"
  ]
};

assertions["should work with complex types 1"] = [
  123,
  123.456,
  "some string",
  true,
  false,
  null,
  {
    "key": "value"
  },
  [
    "element"
  ]
];